프로젝트 매니저 [STEP3] 웡빙

wongbingg opened the pull request:
안녕하세요 제임스 ~! @inwoodev  이번 PR이 많이 늦었네요 ㅜㅜ 내용이 어려워서 구현을 잘 한 것인지 모르겠습니다 😭 
아래 PR메세지에 궁금한 점에대해 남겨 놓았습니다 조언 부탁드립니다!! 🫡

# 프로젝트 매니져 Step3 PR

###### tags: `PR`

## File Tree
```
.
├── ProjectManager
│   ├── swiftlint
│   ├── ProjectManager
│   │   ├── Utility
│   │   │   ├── Observable
│   │   │   ├── NetworkCheck
│   │   │   └── Builder
│   │   │       ├── LabelBuilder
│   │   │       └── StackViewBuilder
│   │   ├── Protocol
│   │   │   └── Coordinator
│   │   ├── Extensions
│   │   │   ├── UITextField+Extension
│   │   │   └── TimeInterval+Extension
│   │   ├── Application
│   │   │   ├── AppDelegate
│   │   │   ├── AppAppearance
│   │   │   └── AppCoordinator
│   │   ├── MainScene
│   │   │   ├── Coordinator
│   │   │   │   └── TodoListCoordinator
│   │   │   ├── ViewModel
│   │   │   │   ├── HeaderViewModel
│   │   │   │   ├── ListCollectionViewModel
│   │   │   │   └── ListCellViewModel
│   │   │   └── View
│   │   │       ├── TodoListViewController
│   │   │       └── ListView
│   │   │           ├── Header
│   │   │           │   └── HeaderView
│   │   │           └── CollectionView
│   │   │               ├── ListCollectionView
│   │   │               ├── ListCollectionView+UIGestureRecognizerDelegate
│   │   │               └── Cell
│   │   │                   └── ListCell
│   │   ├── FormSheetScene
│   │   │   ├── Coordinator
│   │   │   │   └── FormSheetViewCoordinator
│   │   │   ├── ViewModel
│   │   │   │   └── FormSheetViewModel
│   │   │   └── View
│   │   │       ├── FormSheetViewController
│   │   │       └── Template
│   │   │           └── FormSheetTemplateView
│   │   ├── HistoryScene
│   │   │   ├── Coordinator
│   │   │   │   └── HistoryViewCoordinator
│   │   │   ├── ViewModel
│   │   │   │   └── HistoryViewModel
│   │   │   └── View
│   │   │       ├── HistoryViewController
│   │   │       └── HistoryCell
│   │   ├── PopoverScene
│   │   │   ├── Coordinator
│   │   │   │   └── PopoverViewCoordinator
│   │   │   ├── ViewModel
│   │   │   │   └── PopoverViewModel
│   │   │   └── View
│   │   │       └── PopoverViewController
│   │   ├── Model
│   │   │   ├── History
│   │   │   ├── Todo
│   │   │   └── Category
│   │   ├── DataManager
│   │   │   ├── TodoDataManager
│   │   │   ├── RemoteDataManager
│   │   │   └── HistoryManager
│   │   ├── Assets
│   │   ├── LaunchScreen
│   │   └── Info
│   ├── Products
│   ├── Pods
│   └── Frameworks
└── Pods
```

## Coordinator
![](https://i.imgur.com/98QhJPA.png)

## 화면구성
![](https://i.imgur.com/aYpgYgB.png)

## 데이터 구성
### TodoListViewController( 메인 뷰컨트롤러 )

![](https://i.imgur.com/kY1BK2D.png)


### Model
```swift

class Todo: Object {
    @objc dynamic var id: UUID = UUID()
    @objc dynamic var category: String = Category.todo
    @objc dynamic var title: String = ""
    @objc dynamic var body: String = ""
    @objc dynamic var date: Date = Date()
}

```

### TodoDataManager
로컬에 데이터를 저장하는 Realm을 활용하여 CRUD를 구현

### RemoteDataManager
원격저장소에 데이터를 저장하는 Firebase를 활용하여 CRUD 구현



## 기능구현

### Observable
특정 객체를 구독하기 위한 커스텀클래스 
```swift
class Observable<T> {
    var value: T? {
        didSet {
            listeners.forEach {
                $0(value)
            }
        }
    }
    
    init(_ value: T?) {
        self.value = value
    }
    
    private var listeners: [((T?) -> Void)] = []
    
    func bind(_ listener: @escaping (T?) -> Void) {
        listener(value)
        self.listeners.append(listener)
    }
}
```

### - LocalData
Realm 라이브러리를 활용하여 로컬데이터 캐싱작업을 진행![](https://i.imgur.com/wz35Xto.png)

### - RemoteData
FireBase 라이브러리를 활용하여 원격저장소에 저장
앱을 삭제했다가 다시 설치했을 때, 원격저장소의 데이터를 가져와서 동기화시킴![](https://i.imgur.com/KseFw89.png)

### - History
Todo의 추가, 삭제, 이동 을 기록하는 historyManager를 생성
HistoryViewController를 통해 이를 표현
<img src="https://i.imgur.com/rc1WXig.png" width="500">


### - Network State
네트워크가 끊겼을 때, Alert로 알림
<img src="https://i.imgur.com/kIxMzJV.png" width="500">

### - BuilderPattern
- CodeBaseUI 를 작성할 때, 반복되는 공정을 단순화 하고자 빌더패턴을 적용

## ⚒🛠Trouble Shooting

### 1. 클로져에서 [weak self] 캡쳐

습관적으로 클로저 사용 시, [weak self] 를 이용하여 캡쳐리스트를 가져왔는데, 리뷰어분의 질문에 의해 서칭을 한번 해보았다. 좋은 외국 아티클을 찾아 읽어보고 번역하며 내용을 이해해보았다.  [아티클 내용번역](https://hackmd.io/@gFY3eCCiRRKejwfOqnNr2g/BJCnWFbWj)

[weak self] 캡쳐를 해서 self를 사용하려면 optional binding을 해주거나 optional chaining 을 이용해야 하는데, 이 둘의 차이점에 대한 설명이 나와있었다. 

첫번째로, `guard let self = self else {return}` 구문이다. 이렇게 옵셔널 해제를 시켜주면 self가 클로저 scope동안 일시적인 강한참조를 유지하게 된다. 그래서 만약 self(=viewController로 생각해보면) 가 클로저 실행 중간에 dismiss가 된 상황을 생각 해보았을 때, 해제가 되지 않고 클로저 구문이 다 실행된 후에 할당해제가 될 것이다. delayed deallocated 가 되는것이다. 이는 나의 의도에 따라 좋을수도, 나쁠수도있다.

두번째로, `self?.` 옵셔널 체이닝을 이용해 접근하는 것이다. 이 방법으로 self에 접근한다면 위에 말했던 것 처럼 클로저 실행 중간에 self가 dismiss가 된 상황을 생각해보면, 그 시점 이후에 클로저 내부 나머지 코드 중 self로 접근하는 메서드들이 있으면 nil check에서 모두 건너뛰어 질 것이다. 할당해제가 바로 되는 것이다. 이 또한 나의 의도가 어떤지에 따라 좋을수도 나쁠수도 있다. 필요에 따라 두가지 방법을 적절히 사용하는 것이 좋을 것 같다 


### 2. 클로져로 데이터 바인딩 (해결)

뷰 모델이 가지고 있는 클로저 프로퍼티를 어느 뷰에서 바인딩을 할 때, viewModel은 하나이지만 viewModel이 처리해주어야 하는 ListView 는 3개나 된다. 재활용 되는 ListView 한 곳에서 처리해주면 되겠지 싶어 그 안에 바인딩을 해주었는데 왠걸, doneListView에만 바인딩이 되는 것이다..ㅎ 이유는 아마 모두 같은 적용이 되었을 텐데 todoListView, doingListView, doneListView 순서대로 생성을 해주어 마지막에 생성된 doneListView에서 바인딩시켜준 didChangedCount() 가 최종적으로 할당이 된 것이라 생각이 된다.
예시:
``` swift
var number: Int?
number = 1
number = 2
number = 3

//최종은 3이 할당됨
```

**해결방법** : 3개의 ListView를 모두 가지고 있는 ViewController에서 바인딩을 해 주었다. 
**수정** : 위처럼 해결하니 디미터법칙을 위반하게 되었다. 하여 직접 ListCollectionView 에서 바인딩을 해주는데, 할당이 아닌 배열에 추가해주는 방식으로 변경을 하여 3 객체 모두 등록이 되도록 만들었다. 그래서 뷰모델에서도 클로저타입 프로퍼티가 아닌, 클로저배열타입 프로퍼티로 변경해주었다.
```swift
var didChangedCount: [(() -> Void)] = []
var didCreatedTodo: ((Todo) -> Void)?
var didEditedTodo: [(([Todo]) -> Void)] = []
var didMovedTodo: [(() -> Void)] = []
```

클로저 실행 시:
```swift
// in function

didMovedTodo.forEach { $0() }
didChangedCount.forEach { $0() }
```

### 3. 키보드가 텍스트뷰를 가림 (해결)

실기기에서 테스트 결과, 할일의 내용을 작성하는 textView에서 키보드가 textView를 가려 내용이 보이지 않았다. 이 오류해결을 위해 먼저, formSheet 뷰를 scroll뷰로 한번 감싼 뒤, 키보드가 올라와 가리는 높이 만큼 bottomInset을 주어서 작성중인 내용이 가려지지 않도록 구현했다.

### 4. 로컬저장소와 원격저장소의 활용 (해결중...)

로컬저장소를 통하면 네트워크가 없이도 작업이 가능한데, 원격저장소는 왜 필요할까? 라는 의문에서 어떤방식으로 이 둘을 활용할지 케이스 별로 고민
1. 한 어플을 여러개의 기기에서 사용 할 경우 - **해결중**
3. 어플을 지웠다가 다시 설치한 경우 - **해결**
이 경우 로컬에 저장되어 있던 데이터가 모두 지워진다. 따라서, 앱이 설치되고 처음 실행 시, 기존의 원격저장소의 데이터를 불러와주었다.


### 5. 네트워크 연결이 끊겼을 때, 작업을 진행 후 앱을 껐다가 다시 켤 때 (해결)
앱이 켜저있는 상태에서 네트워크가 끊겼다가 다시 연결되면, 원격저장소로 보내지는 작업들이 밀려있다가 실행되는 반면, 중간에 앱이 꺼졌다가 다시 켜지면 원격저장소로 동기화가 되지 않는다.
------------------------수정----------------------------
이후 원격저장소로 가는 어떠한 작업을 수행 시 밀렸던 작업과 함께 수행된다.

### 6. 네트워크 연결상태 확인 (해결중...)

와이파이를 끊으면, .satisfied 상태가 되고 , 와이파이를 연결하면 .unsatisfied 상태가 된다.
정확히 반대로 동작한다. 시뮬레이터의 오류인지 아직 이유를 알 수 없다..!

```swift
// Network MOnitoring 시작
    public func startMonitoring(in viewController: UIViewController) { // 상태값이 반대로 된다 ..
        monitor.start(queue: queue)
        monitor.pathUpdateHandler = { [weak self] path in
            self?.isConnected = path.status == .satisfied
            self?.getConnectionType(path)

            if self?.isConnected == true {
                print("연결됨")
            } else {
                print("연결안됨")
                let alert = UIAlertController(title: "인터넷 연결이 원활하지 않습니다.", message: "Wifi 또는 셀룰러를 활성화 해주세요.", preferredStyle: .alert)
                let confirm = UIAlertAction(title: "확인", style: .default, handler: nil)
                alert.addAction(confirm)
                DispatchQueue.main.async {
                    viewController.present(alert, animated: true)
                }
            }
        }
    }
```

## 궁금한 점

### 1. 로컬데이터와 원격저장소 연동

현재코드상에서는 realm 데이터를 관리해주는 `TodoDataManager` 에서 `RemoteDataManager` 를 인스턴스화 해서 들고있는 상태이고. TodoManager에서 CRUD 가 동작할 때 `RemoteDataManager`가 같이 동작하도록 설정을 해주었습니다. 이 방법으로 로컬과 리모트의 연동이 이루어졌다고 봐도 될지 잘 모르겠습니다. 😭

### 2. 네트워크 연결상태 체크

위 트러블슈팅 6번에 작성한 코드로 네트워크 연결상태를 체크하고 있습니다. 처음 실행할 때는 네트워크가 연결되었는지 확인이 잘 되지만, 실행 중간중간에 와이파이 온오프를 반복 해보면 정반대의 상태로 인식을 합니다 현재 네트워크 상태를 제대로 인식하지 못할 때는 어떻게 처리해주어야 하는지 궁금합니다..!

### 3. 프로젝트의 구조

이전에는 viewModel 하나로 TodoListViewController의 하위에 전달전달을 해 주었는데, realm을 이용하니 데이터 관리가 중앙화 되어 각 view 마다 독립적으로 viewModel을 가질 수 있게 되었습니다. viewController의 부담을 줄여주는 느낌으로 viewModel을 나누어 보았는데 잘 나누어낸 것인지 궁금합니다..!




## 참고 링크
[Apple Article: Displaying transient content in a popover](https://developer.apple.com/documentation/uikit/windows_and_screens/displaying_transient_content_in_a_popover)
[LongTouchGestureRecognizer사용법](http://yoonbumtae.com/?p=4418)
[스크롤 시, 네비게이션 바 자동 숨김처리 비활성화](https://nemecek.be/blog/126/how-to-disable-automatic-transparent-navbar-in-ios-15)
[UIView그림자 만들기](https://babbab2.tistory.com/41)
[키보드가 텍스트뷰를 가릴때, 해결방법](https://seizze.github.io/2019/11/17/iOS에서-키보드에-동적인-스크롤뷰-만들기.html)

wongbingg committed:
refactor : FromSheetView 하나로 create와 edit을 구현하도록  mode로 분기처리

- PageMode 열거형을 만들어 하나의 ViewController가 Creaet와 Edit으로 재사용될 수 있도록 처리
- FormSheetViewModel을 생성하여 ViewController에 노출되지 않아도 되는 프로퍼티를 처리

wongbingg committed:
refactor : 뷰마다의 viewModel 생성

- 기존에 하나의 viewModel을 하위뷰로 계속해서 전달해주던 좋지않은 구조를 가졌음
- 데이터매니저를 싱글톤패턴으로 만들어 데이터관리가 중앙화 됨으로써 취할 수 있는 이득은 각 뷰에 뷰모델을 가질 수 있게 되었다는 점
- Observable 패턴을 이용하여 데이터 매니저가 가지는 배열에 변화가 생길 시 바인딩 시켜준 로직( 뷰 업데이트 )을 실행하도록 설정
- View의 부담을 줄어들게 하기위해 비즈니스로직 분리

wongbingg committed:
refactor : PopoverViewController 수정

- PopoverViewModel 생성

wongbingg committed:
Merge pull request #1 from wongbingg/STEP-2

step3로 이동

wongbingg committed:
docs : ReadMe 업데이트

wongbingg committed:
feat : FireBase 라이브러리 추가

wongbingg committed:
refactor : ModalTransitionScene을 FromSheetScene으로 변경

- 파일명, 폴더명 변경

wongbingg committed:
refactor : 기기별 UI 대응

- iPad 기기별 화면이 적용되도록 고정값에서 비율값으로 변경
- 오토레이아웃 수정

wongbingg committed:
refactor : 불필요한 코드 제거

- 재활용 가능한 메서드를 이용해 불필요한 기존코드 제거

wongbingg committed:
feat : UI요소 생성 시 BuilderPattern 적용

wongbingg committed:
style : "Utility" 폴더 생성

wongbingg committed:
feat : HistoryViewController 생성

- History 모델 생성
- History 모델을 관리하는 HistoryManager 구현
- History 를 보여주는 HistoryCell 생성
- HistoryViewController로 전환하는 Coordinator 생성
- HistoryViewModel 생성

wongbingg committed:
feat : "NetworkCheck" 클래스 구현

- 현재 네트워크 상태를 체크하는 클래스 구현
- 연결이 끊길 시 , Alert알림

wongbingg committed:
refactor : "PopoverView" 크기설정 코드 이동

- "PopoverViewCoordinator" 의 역할이 아닌 코드 분리

wongbingg committed:
refactor : "FormSheetScene" 코드 수정

- 빌더패턴 적용
- 메서드 분리
- 컨벤션 수정

wongbingg committed:
feat : "MainScene" 코드 수정

- HistoryViewController로 이동하는 delegate 메서드 추가

wongbingg committed:
feat : "TimeInterval+Extension" 기능 추가

- timeInterval을 HH:mm:ss 형태의 문자열로 나타내주는 메서드 추가

wongbingg committed:
feat : "TodoDataManager" 코드 수정

- historyManager 인스턴스 추가
- 첫 설치 후 실행시 기존의 원격저장소와 동기화 시키는 setupInitialData() 메서드 추가 (realm과 연동시키지 않음)

wongbingg committed:
refactor : "RemoteDataManager" 코드 수정

- read() 메서드를 원격 저장소의 데이터를 초기데이터로 설정해주는 메서드로 수정

wongbingg committed:
refactor : 컨벤션 수정

wongbingg committed:
docs : ReadMe 업데이트

inwoodev commented on ProjectManager/ProjectManager/FormSheetScene/ViewModel/FormSheetViewModel.swift:
```diff
@@ -0,0 +1,60 @@
+        todo.body = body
+        todo.date = date
+        return todo
+    }
```
뷰모델이 뷰를 알고 있는 구조인거 같아요. 뷰모델이 뷰를 알게 된다면 생길 수 있는 문제점을 조금 더 고민 해 볼 수 있을까요?

wongbingg replied:
말씀주신대로 ViewModel에서 View를 조작해주었던 메서드를 분리 해주었습니다. 

뷰모델이 뷰를 알게 된다면 , 둘다 타입이 클래스라서 순환 사이클이 발생할 수도 있다는 생각을 했습니다. 또한 MVVM 패턴을 사용하는 이유는 뷰가 어떤 특정 모델 플랫폼에 종속되지 않도록 하는 것인데, 위 코드에서는 FormSheetTemplateView가 FormSheetViewModel 에 종속되어 있는 모습인 것 같습니다!

커밋 : https://github.com/yagom-academy/ios-project-manager/pull/201/commits/d77113a65cc39cb2ea41e70d53ae773746e119a4

inwoodev commented on ProjectManager/ProjectManager/DataManager/RemoteDataManager.swift:
```diff
@@ -0,0 +1,87 @@
+                }
+            }
+        }
+    }
```
TodoDataManager는 RemoteDataManager를 인스턴스로 갖고 있는데 RemoteDataManager가 TodoDataManager에 접근을 또 하는 구조는 바람직하지 않은 것 같아요. 상호 참조를 하지 않는 방향으로 리팩토링해 볼 수 있을까요?

inwoodev commented on ProjectManager/ProjectManager/Utility/Builder/StackViewBuilder.swift:
```diff
@@ -0,0 +1,70 @@
+        stackView.backgroundColor = color
+        return self
+    }
+}
```
빌더패턴을 적용해 보셨군요.🙌🙌🙌 이제 체인 형태로 뷰의 속성을 줄 수 있겠네요~ 빌더패턴의 장점은 어떤게 있을까요?

wongbingg replied:
먼저, 가독성 측면에서 이점을 얻을 수 있을 것 같습니다 ! 또한 선택적으로 원하는 옵션들만 설정 해줄 수 있으니 사용성 측면에서도 좋고, 반복되는 생성공정을 좀 더 단순화할 수 있어서 좋은 것 같습니다..!

wongbingg committed:
feat : GoogleService-Info.plist 업로드

inwoodev commented on ProjectManager/ProjectManager/MainScene/Coordinator/TodoListCoordinator.swift:
```diff
@@ -9,43 +9,47 @@ import UIKit
-        let todoListViewController = TodoListViewController(viewModel: viewModel)
+        let todoListViewController = TodoListViewController()
         todoListViewController.delegate = self
-        todoListViewController.todoListView.collectionView?.transitionDelegate = self
-        todoListViewController.doingListView.collectionView?.transitionDelegate = self
-        todoListViewController.doneListView.collectionView?.transitionDelegate = self
+        todoListViewController.todoListView.collectionView.transitionDelegate = self
+        todoListViewController.doingListView.collectionView.transitionDelegate = self
+        todoListViewController.doneListView.collectionView.transitionDelegate = self
         rootViewController = UINavigationController(rootViewController: todoListViewController)
         return rootViewController ?? UIViewController()
     }
```
저번에 남기려다 깜빡한거 같긴하네요. TodoListViewController가 NavigationController에 embed되어야 하는 이유가 궁금합니다~!

wongbingg replied:
다시보니 push ,pop 전환이 일어나지 않으므로 NavigationController가 필요 없다고 판단하여 UINavigationBar 로 대체하였습니다 !

커밋 : https://github.com/yagom-academy/ios-project-manager/pull/201/commits/35768d45acaa5a2d28339aa6aa613f2c9742a411

inwoodev commented on ProjectManager/ProjectManager/MainScene/View/TodoListViewController.swift:
```diff
@@ -81,13 +69,13 @@ final class TodoListViewController: UIViewController {
         doneListView.translatesAutoresizingMaskIntoConstraints = false
         NSLayoutConstraint.activate([
             todoListView.widthAnchor.constraint(
-                equalToConstant: 356
+                equalToConstant: view.bounds.width/3 - 4
             ),
             doingListView.widthAnchor.constraint(
-                equalToConstant: 356
+                equalToConstant: view.bounds.width/3 - 4
             ),
             doneListView.widthAnchor.constraint(
-                equalToConstant: 356
+                equalToConstant: view.bounds.width/3 - 4
```
하드코딩보다는 더 나은 접근인거 같아요. 아실거 같긴한데 퀴즈를 하나 내볼게요. 리팩토링 전과 후 중 어떤 방식이 퍼포먼스적인 측면에서 더 나은지 알고 계신가요?

wongbingg replied:
view.bounds 로 접근하므로 다른 기기에서도 화면비율을 유지할 수 있다는 장점이 있을 것 같습니다 ! 
제가 퀴즈의 의도를 파악한 것이 맞을까요? 아니면 메모리 관점에서 바라봐야 할까요?

inwoodev replied:
분명 답글을 남겼었는데..지워진거 같아요..무튼 요지는 퍼포먼스, 즉 뷰를 그리는 속도 측면에서 여쭤본거였어요!

inwoodev commented on ProjectManager/ProjectManager/MainScene/ViewModel/HeaderViewModel.swift:
```diff
@@ -0,0 +1,52 @@
+            view.countLabel.text = " \(doneListCount) "
+        default:
+            return
+        }
```
데이터 흐름이 다음과 같이 끊기는 느낌을 받았어요
1. TaskDataManger -> HeaderViewModel
2. HeaderViewModel -> HeaderView
그리고 이 과정에서 ViewModel은 HeaderView를 알고 있는 상황이 되네요.

mvvm에서 제공하는 장점이 view는 ViewModel을 아는데 viewModel은 view를 몰라요.
힌트를 하나 드리자면 만약 기획자가 HeaderView가 아닌 collectionView에 해당 작업을 바인딩하라고 하면 어떡하죠?

wongbingg replied:
위 코멘트에서와 같이 view를 종속시키지 않기 위해 분리 해주었습니다 ! https://github.com/yagom-academy/ios-project-manager/pull/201/commits/869ef3e927a0ff04711fdbe89b573db21764e97e
또한 데이터의 흐름이 끊겨 보인다고 하신 부분을 다시한번 수정해보았습니다 !https://github.com/yagom-academy/ios-project-manager/pull/201/commits/2f899acdb5988e059fb6770e58f5226988c23b61

inwoodev commented on ProjectManager/ProjectManager/MainScene/ViewModel/ListCollectionViewModel.swift:
```diff
@@ -0,0 +1,43 @@
+    
+    func configure(_ view: ListCollectionView) {
+        
+    }
```
불필요한 코드는 포함시키지 말아주세요!

wongbingg replied:
앗.. 이런 실수를 .. ! 😢😢 좀더 꼼꼼하게 살펴보겠습니다 !! 🫡

inwoodev commented on ProjectManager/ProjectManager/FormSheetScene/Coordinator/FormSheetViewCoordinator.swift:
```diff
@@ -0,0 +1,32 @@
+    }
+    
+    func start() -> UIViewController {
+        let formSheetVM = FormSheetViewModel(mode: mode,
+                                        category: category,
+                                        index: index)
+        let formSheetVC = FormSheetViewController(viewModel: formSheetVM)
+        let navCon = UINavigationController(
+            rootViewController: formSheetVC
+        )
+        navCon.modalPresentationStyle = .formSheet
+        return navCon
+    }
```
네비게이션컨트롤러에 embed해야하는 이유가 있을까요?

wongbingg replied:
https://github.com/yagom-academy/ios-project-manager/pull/201#discussion_r980046277

위 코멘트에 대한 수정사항과 마찬가지로 불필요하게 navigationController를 사용한 부분 모두 UINavigationBar로 대체 해주었습니다!

inwoodev commented on ProjectManager/ProjectManager/DataManager/TodoDataManager.swift:
```diff
@@ -5,60 +5,115 @@
 import RealmSwift
 
 final class TodoDataManager {
+    private let firebaseManager = RemoteDataManager()
+    private let historyManager = HistoryManager()
```
높은 결합도를 가지게 되는데 이때문에 생길 수 있는 문제가 있을까요?

inwoodev commented on ProjectManager/ProjectManager/DataManager/TodoDataManager.swift:
```diff
@@ -5,60 +5,115 @@
     private let realm = try? Realm()
+    static let shared = TodoDataManager()
+    
+    var todoList: Observable<[Todo]> = Observable([])
+    var doingList: Observable<[Todo]> = Observable([])
+    var doneList: Observable<[Todo]> = Observable([])
```
todoDataManager가 이 요소들을 가지고 있는게 적합한지는 생각 해 볼 만한 문제인거 같아요. 이 부분은 싱글톤 객체와도 연결되니 같이 고민 해 보시는 것도 좋을거 같습니다~

inwoodev commented on ProjectManager/ProjectManager/Utility/NetworkCheck.swift:
```diff
@@ -0,0 +1,72 @@
+                    handler: nil
+                )
+                alert.addAction(confirm)
+                DispatchQueue.main.async {
+                    viewController.present(alert, animated: true)
+                }
```
Solid의 단일책임원칙에 위배되는 부분인거 같은데 어떻게 생각하시나요?

inwoodev commented on ProjectManager/ProjectManager/MainScene/Coordinator/TodoListCoordinator.swift:
```diff
@@ -9,43 +9,47 @@ import UIKit
-        todoListViewController.todoListView.collectionView?.transitionDelegate = self
-        todoListViewController.doingListView.collectionView?.transitionDelegate = self
-        todoListViewController.doneListView.collectionView?.transitionDelegate = self
+        todoListViewController.todoListView.collectionView.transitionDelegate = self
+        todoListViewController.doingListView.collectionView.transitionDelegate = self
+        todoListViewController.doneListView.collectionView.transitionDelegate = self
```
이 부분은 같이 한 번 더 공부해 봅시다 :)

inwoodev replied:
좀 더 생각 해 본 결과 delegate를 주입시키는 방법으로도 해결 해 볼 수 있을 거 같아요~

inwoodev commented on ProjectManager/ProjectManager/MainScene/View/ListView/CollectionView/ListCollectionView.swift:
```diff
@@ -96,43 +94,18 @@ extension ListCollectionView {
         snapshot.deleteItems([todo])
         todoDataSource?.apply(snapshot, animatingDifferences: true)
-        viewModel.delete(todo: todo)
+        TodoDataManager.shared.delete(todo)
```
TodoDataManager를 여기서 호출하는게 적절한지는 조금 더 고민 해 보면 좋을거 같습니다~

inwoodev commented on ProjectManager/ProjectManager/MainScene/ViewModel/ListCollectionViewModel.swift:
```diff
@@ -0,0 +1,43 @@
+
+final class ListCollectionViewModel {
+    var category: String
+    var list: [Todo] = [] {
+        didSet {
+            didUpdatedList?(list)
+        }
+    }
+    var didUpdatedList: (([Todo]) -> Void)?
+    init(category: String) {
+        self.category = category
+        bind(at: category)
+    }
+    
+    func bind(at category: String) {
+        switch category {
+        case Category.todo:
+            TodoDataManager.shared.todoList.bind { [weak self] (list) in
+                self?.list = list ?? []
+            }
+        case Category.doing:
+            TodoDataManager.shared.doingList.bind { [weak self] (list) in
+                self?.list = list ?? []
+            }
+        case Category.done:
+            TodoDataManager.shared.doneList.bind { [weak self] (list) in
+                self?.list = list ?? []
+            }
+        default:
+            return
+        }
+    }
```
싱글톤에서 바인딩 된게 뷰모델로 오고 뷰모델로 온게 `didSet`에서 vc로 가는 구조네요.
싱글톤에서 vc로 바인딩 될 때 뷰모델이 이 다리역할을 하도록 리팩토링 하는건 어떨까요?

dataManager -> ViewModel -> VC
dataManager __ViewModel__> VC

inwoodev reviewed (changes requested):
> 1. 로컬데이터와 원격저장소 연동
현재코드상에서는 realm 데이터를 관리해주는 TodoDataManager 에서 RemoteDataManager 를 인스턴스화 해서 들고있는 상태이고. TodoManager에서 CRUD 가 동작할 때 RemoteDataManager가 같이 동작하도록 설정을 해주었습니다. 이 방법으로 로컬과 리모트의 연동이 이루어졌다고 봐도 될지 잘 모르겠습니다. 😭

일단 코드를 봤을 때는 로컬과 리모트의 연동이 이루어졌다고 봐도 무방한거 같아요!. TodoDataManager에서 동기적으로 realm에 CRUD를 하면서 realm에도 CRUD를 동시에 진행하니까요.

---

> 2. 네트워크 연결상태 체크
위 트러블슈팅 6번에 작성한 코드로 네트워크 연결상태를 체크하고 있습니다. 처음 실행할 때는 네트워크가 연결되었는지 확인이 잘 되지만, 실행 중간중간에 와이파이 온오프를 반복 해보면 정반대의 상태로 인식을 합니다 현재 네트워크 상태를 제대로 인식하지 못할 때는 어떻게 처리해주어야 하는지 궁금합니다..!

이 부분은 dm으로 얘기한거 같이 실 기기로 테스트 한 번 해보시면 좋을거 같습니다.

---

> 3. 프로젝트의 구조
이전에는 viewModel 하나로 TodoListViewController의 하위에 전달전달을 해 주었는데, realm을 이용하니 데이터 관리가 중앙화 되어 각 view 마다 독립적으로 viewModel을 가질 수 있게 되었습니다. viewController의 부담을 줄여주는 느낌으로 viewModel을 나누어 보았는데 잘 나누어낸 것인지 궁금합니다..!

일단 view마다 viewModel을 갖게 하여 MassiveViewController를 방지하는 하였기 때문에 코드 리뷰하기가 더 수월해졌기에 유지보수하기 더 적합하게 된거 같아서 저는 긍정적인거 같습니다. 다만 TodoDataManager를 호출하는 객체들이 어떤 객체여야 하는지는 조금 더 생각 해 볼만한 문제일거 같아요.

그리고 저희 디스코드에서 얘기 나눴던 부분을 조금 더 얘기 해 볼게요. 일단 TodoDataManager가 Observable배열들을 가지고 있는게 적합한지 아닌지부터 좀 더 고민 해 봤으면 좋겠어요. 그 이후에 이제 해당 객체가 싱글톤일 필요가 있는지. 다른 말로는 싱글 아이덴티티를 갖고 있을 필요가 있는지도 생각 해 보면 좋겠어요.

binding같은 경우도 조금 더 step을 줄일 수도 있을거 같아서 이 부분도 숙제로 남겨봅니다. comment에도 조금씩 힌트 남겨두었으니까 같이 한 번 더 고민 해 보시면 좋을거 같아요. 

더 궁금한 점 있으면 언제든 dm 주세요~!

wongbingg committed:
refactor : "NetworkCheck" 클래스 수정

- 메서드 분리

wongbingg committed:
refactor : BuilderPattern 추가 및 수정

- 모든 UI요소 생성로직에 BuilderPattern 구현
- 공통되는 메서드를 UseAutoLayout, UseLayer 두개의 프로토콜로 분리

wongbingg committed:
refactor : "HistoryViewController" 를 "HistoryTableViewController"로 수정

- 많은 데이터가 들어올 때, 스크롤을 통해 볼 수 있도록 테이블뷰로 변경

wongbingg committed:
refactor : "ListCollectionView" 수정

- 데이터 바인딩 방식 변경
- 데이터 이동의 경우 모든 배열을 받아와 화면 update 진행
- 반면 (추가, 삭제)의 경우 추가/ 삭제된 아이템 하나만 받아와서 화면 update 진행
   [추가/ 삭제 시 "TodoDataManager" 에서 클로저를 통해 전달]

wongbingg committed:
refactor : "RemoteDataManager" 수정

- 메서드 분리

wongbingg committed:
refactor : "HistoryViewCoordinator" 수정

- popover Transition의 기준점을 sourceView가 아닌 barButtonItem으로 번경

wongbingg committed:
chore : 기타 수정사항

- final 처리
- popoverView 의 화살표 방향 변경

wongbingg committed:
refactor : "FormSheetViewModel" 에서 View 분리

- 뷰모델이 뷰를 알고있는 구조 지양

wongbingg committed:
refactor : UINavigationController 를 UINavigationBar 로 수정

- 불필요한 UINavigationController 를 제거하고, UINavigationBar 로 UI를 만들어 적용

wongbingg committed:
refactor : "HeaderViewModel" 에서 View 분리

- 뷰모델이 뷰를 알고있는 구조 지양

wongbingg committed:
refactor : "HeaderViewModel"  데이터 바인딩 수정

- 2번 거쳐가던 데이터 흐름을 1번 으로 수정

wongbingg committed:
refactor : 불필요한 ViewModel 제거

wongbingg committed:
refactor : "TodoDataManger" 수정

- 불필요한 배열 제거
- 불필요한 메서드 제거

wongbingg committed:
refactor : "ListView" 에서 category 상수 전달 끊기

- ListView 의 하위뷰인 HeaderView와 ListCollectionView 에 ViewModel을 주입
- category를 하위 뷰로 직접 넘길 필요 없고 viewModel만 넘기도록

wongbingg committed:
refactor : : "HeaderView" 데이터바인딩 수정

- 바인딩의 단계를 줄인 형태로 수정

wongbingg committed:
refactor : "ListCollectionView" 데이터 바인딩 수정

- 바인딩의 단계를 줄인 형태로 수정

wongbingg committed:
refactor : "TodoListCoordinator" delegate 채택 주입

- 하위 뷰로 delegate를 전달하여 한번에 채택하는 방법으로 변경

wongbingg committed:
refactor : "TodoDataManager" 와 "RemoteDataManager" 결합도 최소화

- 결합도를 낮추기 위해 delegate 패턴으로 연결
- "RemoteDataManager" 가 delegate를 채택할 때, 부자연스러움
- etc : Todo 모델 convenience init 추가

inwoodev commented on ProjectManager/ProjectManager/MainScene/View/TodoListViewController.swift:
```diff
@@ -33,6 +33,18 @@ final class TodoListViewController: UIViewController {
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
```
주입하는 구간이 늘어나서 boilerplate코드가 늘어나는 점은 coordinator패턴의 단점 중 하나라 생각됩니다. 사실 이런 불편함을 감수하더라도 coordinator pattern을 사용하는지 적합한지 한번 쯤 고민해도 좋을 타이밍이네요~
